# Starter Code

Here, we briefly introduce the data reading procedures and the data structure.

## Read Data

In [17]:
import h5py

# You may need to change the following paths according to your file structure.
path_image = '../datasets/training_set/train_sample3_vol0/img_zyx_2217-2617_4038-4448_6335-6735.h5'
path_label = '../datasets/training_set/train_sample3_vol0/syns_zyx_2217-2617_4038-4448_6335-6735.h5'

### Image

In [18]:
f_image = h5py.File(path_image, 'r')
print(f_image.keys())

<KeysViewHDF5 ['main', 'voxel_offset', 'voxel_size']>


The key ```main``` indicates the image data.

In [19]:
image = f_image['main'][:]
print(image.shape)
f_image.close()

(400, 410, 400)


### Label

In [20]:
f_label = h5py.File(path_label, 'r')
print(f_label.keys())

<KeysViewHDF5 ['post', 'post_users', 'pre', 'pre_confidence', 'pre_users', 'resolution']>


The key ```pre``` indicates labels of pre-synapses and the key ```post``` indicates labels of post-synapses.

In [21]:
pre = f_label['pre'][:]
post = f_label['post'][:]
f_label.close()

In [22]:
print(pre.shape)
print(pre[:10, :]) # We take 10 pre-synapses for example.

(295, 3)
[[2234 4047 6353]
 [2223 4051 6450]
 [2239 4040 6463]
 [2228 4077 6612]
 [2222 4091 6695]
 [2224 4091 6682]
 [2229 4100 6341]
 [2233 4198 6450]
 [2219 4199 6463]
 [2221 4168 6464]]


Each row contains a 3D coordinate for a pre-synapse. There are 295 pre-synapses in this sample.

In [23]:
print(post.shape)
print(post[:10, :]) # We take 10 post-synapses for example.

(880, 4)
[[  25 2236 4051 6360]
 [   0 2239 4037 6364]
 [   6 2233 4084 6336]
 [   6 2233 4091 6351]
 [   2 2221 4038 6454]
 [   1 2227 4046 6462]
 [   1 2223 4036 6455]
 [   1 2206 4068 6459]
 [   2 2229 4041 6464]
 [   3 2239 4063 6604]]


In each row, the 2nd to the 4th columns contain the coordinate of a post-synapse. The 1st column is the ID (row index in ```pre``` data) of a pre-synapse to which this post-synapse is connected. There are 880 post-synapses in this sample.

**You need to offset the labels!**

In [24]:
offset_zyx = path_label.split('/')[-1].split('_')
offset_z = int(offset_zyx[2].split('-')[0])
offset_y = int(offset_zyx[3].split('-')[0])
offset_x = int(offset_zyx[4].split('-')[0])
pre = pre - [offset_z, offset_y, offset_x]
post[:, 1:] = post[:, 1:] - [offset_z, offset_y, offset_x]